solver/src/solver/matcha_solver.rs

```diff
@@ -0,0 +1,312 @@
+    }
+}
+
+#[async_trait::async_trait]
+impl<F: AllowanceFetching> SingleOrderSolving for MatchaSolver<F> {
+    async fn settle_order(&self, order: LimitOrder) -> Result<Option<Settlement>> {
+        let (swap, mut settlement) = match order.kind {
+            OrderKind::Sell => {
+                let query = SwapQuery {
+                    sell_token: order.sell_token,
+                    buy_token: order.buy_token,
+                    sell_amount: Some(order.sell_amount),
+                    buy_amount: None,
+                    slippage_percentage: Slippage::basis_points(STANDARD_MATCHA_SLIPPAGE_BPS)
+                        .unwrap(),
+                    skip_validation: Some(true),
+                };
+
+                tracing::debug!("querying Matcha swap api with {:?}", query);
+                let swap = self.client.get_swap(query).await?;
+                tracing::debug!("proposed Matcha swap is {:?}", swap);
+
+                ensure!(
+                    swap.buy_amount >= order.buy_amount,
+                    "order limit price not respected",
+                );
+                let settlement = Settlement::new(hashmap! {
+                    order.sell_token => swap.sell_amount,
+                    order.buy_token => swap.buy_amount,
+                });
+                (swap, settlement)
+            }
+            OrderKind::Buy => {
+                let query = SwapQuery {
+                    sell_token: order.sell_token,
+                    buy_token: order.buy_token,
+                    sell_amount: None,
+                    buy_amount: Some(order.buy_amount),
+                    slippage_percentage: Slippage::basis_points(STANDARD_MATCHA_SLIPPAGE_BPS)
+                        .unwrap(),
+                    // From the api documentation:
+                    // SlippagePercentage(Optional): The maximum acceptable slippage in % of the buyToken amount if sellAmount is provided, the maximum acceptable slippage in % of the sellAmount amount if buyAmount is provided. This parameter will change over time with market conditions.
+                    // => Hence, allegedly, we don't need to build in buffers ourselves. Though the sell amount is not adjusted, if slippage is changed for buy order requests.
+                    // Todo: Continue discussion with the 0x-team
+                    skip_validation: Some(true),
+                };
+
+                tracing::debug!("querying Matcha swap api with {:?}", query);
+                let swap = self.client.get_swap(query).await?;
+                tracing::debug!("proposed Matcha swap is {:?}", swap);
+
+                ensure!(
+                    swap.sell_amount <= order.sell_amount,
+                    "order limit price not respected",
+                );
+                let settlement = Settlement::new(hashmap! {
+                    order.sell_token => swap.sell_amount,
+                    order.buy_token => swap.buy_amount,
+                });
```

Comment:
This block seems very similar to the one above for sell orders. Although the inequality is different, I wonder if it could somehow be pulled out and only exist once.

Reply:
Yeah, you are right. though I could not come up with a good solution. The tricky part is that the ensure in the middle of the code is different

Reply:
Since the only difference is the ensure statement, it might be less code to match on the order type again outside this block just for the ensure statement. Or even make a helper function for the inequality checking that the ensure statement calls into. Just a thought might look a lot less repetitive and then you could have an ensure(limit price respected) which looks kinda cool.